File: src/Core/Models/TaskResult.cs
namespace CyberPatriotAutomation.Core.Models;

/// <summary>
/// Represents the result of executing a remediation task
/// </summary>
public class TaskResult
{
    public string TaskName { get; set; } = string.Empty;
    public bool Success { get; set; }
    public string Message { get; set; } = string.Empty;
    public string? ErrorDetails { get; set; }
    public DateTime ExecutedAt { get; set; } = DateTime.Now;

    /// <summary>
    /// Number of items/actions attempted in this task
    /// </summary>
    public int ItemsAttempted { get; set; }

    /// <summary>
    /// Number of items/actions that succeeded
    /// </summary>
    public int ItemsSucceeded { get; set; }

    /// <summary>
    /// Number of items that were skipped (already compliant or N/A)
    /// </summary>
    public int ItemsSkipped { get; set; }

    /// <summary>
    /// Confidence level (0-100) based on verification results
    /// </summary>
    public int ConfidencePercent { get; set; } = 100;

    /// <summary>
    /// Whether verification passed after execution
    /// </summary>
    public bool Verified { get; set; }

    /// <summary>
    /// Calculates completion rate as a percentage
    /// </summary>
    /// <remarks>
    /// When a task reports no per-item counts, fall back to whether the task
    /// itself succeeded. Returning a flat 100% in that case reported full
    /// completion for tasks that had failed outright - and since no task
    /// populates <see cref="ItemsAttempted"/>, that was every task.
    /// </remarks>
    public double CompletionRate =>
        ItemsAttempted > 0 ? (double)(ItemsSucceeded + ItemsSkipped) / ItemsAttempted * 100
        : Success ? 100
        : 0;
}

File: rust/src/models/task_result.rs
use chrono::{DateTime, Local};

/// Represents the result of executing a remediation task.
#[derive(Debug, Clone)]
pub struct TaskResult {
    pub task_name: String,
    pub success: bool,
    pub message: String,
    pub error_details: Option<String>,
    pub executed_at: DateTime<Local>,

    /// Number of items/actions attempted in this task.
    pub items_attempted: i32,
    /// Number of items/actions that succeeded.
    pub items_succeeded: i32,
    /// Number of items that were skipped (already compliant or N/A).
    pub items_skipped: i32,

    /// Confidence level (0-100) based on verification results.
    pub confidence_percent: i32,
    /// Whether verification passed after execution.
    pub verified: bool,
}

impl Default for TaskResult {
    fn default() -> Self {
        Self {
            task_name: String::new(),
            success: false,
            message: String::new(),
            error_details: None,
            executed_at: Local::now(),
            items_attempted: 0,
            items_succeeded: 0,
            items_skipped: 0,
            confidence_percent: 100,
            verified: false,
        }
    }
}

impl TaskResult {
    /// Calculates completion rate as a percentage.
    ///
    /// When a task does not report per-item counts, fall back to whether the
    /// task itself succeeded. Returning a flat 100% in that case - as this
    /// previously did - reported full completion for tasks that had failed
    /// outright, since no task populates `items_attempted`.
    pub fn completion_rate(&self) -> f64 {
        if self.items_attempted > 0 {
            (self.items_succeeded + self.items_skipped) as f64 / self.items_attempted as f64 * 100.0
        } else if self.success {
            100.0
        } else {
            0.0
        }
    }
}
